### ISSUE TITLE
Conflict with ncdump and ncgen between hdf and netcdf-c packages.

### ISSUE THREAD

**Issue by srherbener** [JCSDA/spack-stack#205]
When I try to run `ncdump netcdf-file` I get the following error message:
```
*** ncdump: ncopen failed on ioda/test/testoutput/amsua_n19_obs_2018041500_m_put_db_0000.nc4
```

This is being caused by both netcdf-c and hdf packages building ncdump executables, and the wrong one is being picked up. The ncdump from the netcdf-c is the one that should be used, but the ncdump from the hdf package is the one that is found.

Same thing for the ncgen executable.

**Comment by srherbener** [comment]
@kgerheiser, @climbfuji I just added this issue, and I can't seem to add an estimate nor the epic it belongs to. Am I doing this wrong, or missing permissions somewhere? Thanks!

**Comment by climbfuji** [comment]
> @kgerheiser, @climbfuji I just added this issue, and I can't seem to add an estimate nor the epic it belongs to. Am I doing this wrong, or missing permissions somewhere? Thanks!

I figured out what the problem is. The hdf4-4.2.15 install also creates a binary called `ncdump`, and because of the order of how the modules are loaded it finds that one:
```
heinzell@JCSDA-L-18146:~/scratch/ufs-run/control_c48 [brew-x86_64]> which ncdump
/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/hdf-4.2.15-rpy26yf/bin/ncdump
```

The netcdf-c install is pretty much complete:
```
heinzell@JCSDA-L-18146:~/scratch/ufs-run/control_c48 [brew-x86_64]> nc-config --all

This netCDF 4.8.1 has been built with the following features:

  --cc            -> /Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/openmpi-4.1.3-hldiips/bin/mpicc
  --cflags        -> -I/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-c-4.8.1-ygcbmzc/include -I/usr/local/Cellar/curl/7.83.0/include -I/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/hdf5-1.12.1-jn7p2sk/include -I/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/parallel-netcdf-1.12.2-zworqap/include
  --libs          -> -L/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-c-4.8.1-ygcbmzc/lib -lnetcdf
  --static        -> -lpnetcdf -lhdf5_hl -lhdf5 -lm -lz -lcurl

  --has-c++       -> no
  --cxx           ->

  --has-c++4      -> yes
  --cxx4          -> /Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/spack/lib/spack/env/clang/clang++
  --cxx4flags     -> -I/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-cxx4-4.3.1-jxyenvz/include
  --cxx4libs      -> -L/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-cxx4-4.3.1-jxyenvz/lib -lnetcdf_c++4 -lnetcdf

  --has-fortran   -> yes
  --fc            -> /Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/spack/lib/spack/env/clang/gfortran
  --fflags        -> -I/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-fortran-4.5.4-5cv747y/include -I/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-fortran-4.5.4-5cv747y/include
  --flibs         -> -L/Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-fortran-4.5.4-5cv747y/lib -lnetcdff -lnetcdf -lnetcdf -lm
  --has-f90       ->
  --has-f03       -> yes

  --has-dap       -> yes
  --has-dap2      -> yes
  --has-dap4      -> yes
  --has-nc2       -> yes
  --has-nc4       -> yes
  --has-hdf5      -> yes
  --has-hdf4      -> no
  --has-logging   -> no
  --has-pnetcdf   -> yes
  --has-szlib     -> no
  --has-cdf5      -> yes
  --has-parallel4 -> yes
  --has-parallel  -> yes
  --has-nczarr    -> yes

  --prefix        -> /Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-c-4.8.1-ygcbmzc
  --includedir    -> /Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-c-4.8.1-ygcbmzc/include
  --libdir        -> /Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/netcdf-c-4.8.1-ygcbmzc/lib
  --version       -> netCDF 4.8.1
```

**Comment by climbfuji** [comment]
This means that you probably want to provide more info in the description and title that this has to do with a conflict with hdf4 binaries. Definitely needs to be fixed in the release, I'll add the corresponding labels. Thinking about a solution ...

**Comment by climbfuji** [comment]
@srherbener The same problem for `ncgen`:
```
heinzell@JCSDA-L-18146:~/scratch/ufs-run/control_c48 [brew-x86_64]> ls -1 /Users/heinzell/work/spack-stack/spack-stack-1.0.0-rc1/envs/skylab-1.0.0/install/apple-clang/13.1.6/hdf-4.2.15-rpy26yf/bin/
gif2hdf
h4cc
h4redeploy
hdf24to8
hdf2gif
hdf2jpeg
hdf8to24
hdfcomp
hdfed
hdfimport
hdfls
hdfpack
hdftopal
hdftor8
hdfunpac
hdiff
hdp
hrepack
jpeg2hdf
ncdump
ncgen
paltohdf
r8tohdf
ristosds
vmake
vshow
```

**Comment by edwardhartnett** [comment]
Ok this is a tricky one actually. Lots of installers have tripped over this.

HDF4 made the unwise decision to include their own copy of the netcdf API and tools.

Problem is, it's a very old version of netCDF (long before netcdf4). And it's never been updated. So it will only work on netcdf classic files.

The answer is that HDF4 must be built with --disable-netcdf. This should be done by spack always.

Ironically the opposite situation works well - netCDF has a --enable-hdf4 option which works well for HDF4 SD files. This works because instead of reimplementing HDF4, netCDF uses the HDF4 library and just calls it.

**Comment by climbfuji** [comment]
Well, that is a perfect solution, @edwardhartnett - thanks. We'll make this change in release candidate 2.

**Comment by climbfuji** [comment]
Turns out that this was my fault, because I explicitly enable the netcdf interface for hdf when I added the package in spack-stack. Will be fixed in #208.

**Comment by climbfuji** [comment]
Closed via https://github.com/NOAA-EMC/spack-stack/pull/208
